vendor/github.com/mongodb/mongo-tools-common/db/buffered_bulk.go

```diff
@@ -12,85 +12,99 @@ import (
 
 	"go.mongodb.org/mongo-driver/bson"
 	"go.mongodb.org/mongo-driver/mongo"
-	mopt "go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
 // BufferedBulkInserter implements a bufio.Writer-like design for queuing up
 // documents and inserting them in bulk when the given doc limit (or max
 // message size) is reached. Must be flushed at the end to ensure that all
 // documents are written.
 type BufferedBulkInserter struct {
-	collection               *mongo.Collection
-	documents                []interface{}
-	docLimit                 int
-	byteCount                int
-	docCount                 int
-	unordered                bool
-	bypassDocumentValidation bool
+	collection    *mongo.Collection
+	writeModels   []mongo.WriteModel
+	docLimit      int
+	docCount      int
+	bulkWriteOpts *options.BulkWriteOptions
+	upsert        bool
 }
 
 // NewBufferedBulkInserter returns an initialized BufferedBulkInserter
 // for writing.
 func NewBufferedBulkInserter(collection *mongo.Collection, docLimit int, unordered bool) *BufferedBulkInserter {
 	bb := &BufferedBulkInserter{
-		collection:      collection,
-		unordered:       unordered,
-		docLimit:        docLimit,
+		collection:    collection,
+		bulkWriteOpts: options.BulkWrite().SetOrdered(!unordered),
+		docLimit:      docLimit,
 	}
 	bb.resetBulk()
 	return bb
 }
 
 func (bb *BufferedBulkInserter) Unordered() {
-	bb.unordered = true
+	bb.bulkWriteOpts.SetOrdered(false)
 }
 
 func (bb *BufferedBulkInserter) SetBypassDocumentValidation(bypass bool) {
-	bb.bypassDocumentValidation = bypass
+	bb.bulkWriteOpts.SetBypassDocumentValidation(bypass)
+}
+
+func (bb *BufferedBulkInserter) SetUpsert(upsert bool) {
+	bb.upsert = upsert
 }
 
 // throw away the old bulk and init a new one
 func (bb *BufferedBulkInserter) resetBulk() {
-	bb.documents = make([]interface{}, 0, bb.docLimit)
-	bb.byteCount = 0
+	bb.writeModels = make([]mongo.WriteModel, 0, bb.docLimit)
```

Comment:
Since we're never resizing this slice, I think we can reuse the memory by doing something like `bb.writeModels = bb.writeModels[:0]`

Reply:
Done